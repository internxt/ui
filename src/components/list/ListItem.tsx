import { LegacyRef, useEffect, useRef, useState } from 'react';
import { DotsThree } from '@phosphor-icons/react';
import ContextMenu from '../contextMenu/ContextMenu';
import useHotkeys from '../../hooks/useHotKeys';
import Checkbox from '../checkbox/Checkbox';
import { MenuItemsType } from '../menu/Menu';

interface ListItemProps<T> {
  item: T;
  listIndex: number;
  itemComposition: Array<(props: T) => JSX.Element>;
  selected: boolean;
  isOpen: boolean;
  columnsWidth: Array<string>;
  onClose: () => void;
  onSelectedChanged: (value: boolean) => void;
  onDoubleClick?: () => void;
  onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onClickContextMenu?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onThreeDotsButtonPressed?: (e: React.MouseEvent<HTMLButtonElement>) => void;
  menu?: MenuItemsType<T>;
  disableItemCompositionStyles?: boolean;
  onMouseEnter?: () => void;
  onMouseLeave?: () => void;
  genericEnterKey: () => void;
}
const TOP_MIN_HEIGHT = 500;

const ListItem = <T extends { id: number }>({
  item,
  listIndex,
  itemComposition,
  selected,
  isOpen,
  columnsWidth,
  onClose,
  onSelectedChanged,
  onDoubleClick,
  onClick,
  onClickContextMenu,
  onThreeDotsButtonPressed,
  disableItemCompositionStyles,
  menu,
  onMouseEnter,
  onMouseLeave,
  genericEnterKey,
}: ListItemProps<T>): JSX.Element => {
  const menuButtonRef = useRef<HTMLButtonElement | undefined>();
  const rootWrapperRef = useRef<HTMLDivElement | null>(null);
  const menuItemsRef = useRef<HTMLDivElement | null>(null);

  const [openedFromRightClick, setOpenedFromRightClick] = useState(false);
  const [posX, setPosX] = useState(0);
  const [posY, setPosY] = useState(0);
  const [dimensions, setDimensions] = useState({ width: 0, height: 0 });
  const [isContextMenuCutOff, setIsContextMenuCutOff] = useState(false);

  useEffect(() => {
    if (
      menuItemsRef.current &&
      (menuItemsRef.current.offsetHeight !== dimensions.height ||
        menuItemsRef.current?.offsetWidth !== dimensions.width)
    )
      setDimensions({
        width: menuItemsRef.current.offsetWidth,
        height: menuItemsRef.current.offsetHeight,
      });
  }, [dimensions.height, dimensions.width]);

  useEffect(() => {
    if (!openedFromRightClick) handleOpenPosition();

    if (!open) {
      setOpenedFromRightClick(false);
      setPosX(0);
      setPosY(0);
    }
  }, [open]);

  function handleOpenPosition() {
    const element = menuButtonRef.current;
    const contextMenuHeight = menuItemsRef?.current?.offsetHeight || 300;
    if (!element) return;
    if (!contextMenuHeight) return;

    const { bottom } = element.getBoundingClientRect();
    const windowHeight = window.innerHeight;

    const isContextCutOff = bottom + contextMenuHeight > windowHeight;
    setIsContextMenuCutOff(isContextCutOff);
  }

  const handleContextMenuClick = (event: React.MouseEvent<HTMLDivElement>) => {
    event.preventDefault();

    onClickContextMenu?.(event);
    const childWidth = menuItemsRef?.current?.offsetWidth || 240;
    const childHeight = menuItemsRef?.current?.offsetHeight || 300;
    const wrapperRect = rootWrapperRef?.current?.getBoundingClientRect();
    const { innerWidth, innerHeight } = window;
    let x = event.clientX - (wrapperRect?.left || 0);
    let y = event.clientY - (wrapperRect?.top || 0);

    if (event.clientX + childWidth > innerWidth) {
      x = x - childWidth;
    }

    if (event.clientY + childHeight > innerHeight && event.clientY > TOP_MIN_HEIGHT) {
      y = y - childHeight;
    }

    setPosX(x);
    setPosY(y);
    setOpenedFromRightClick(true);
  };

  const handleContextMenuDotsButton = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();

    onThreeDotsButtonPressed?.(event);
    setOpenedFromRightClick(false);
  };

  const handleClick = (event: React.MouseEvent<HTMLDivElement>) => {
    onClick?.(event);
    openedFromRightClick && handleMenuClose();
  };

  const handleMenuClose = () => {
    onClose();
  };

  const handleEnterKey = () => {
    if (!isOpen) {
      genericEnterKey();
    }
  };

  useHotkeys(
    {
      r: handleMenuClose,
      backspace: handleMenuClose,
      enter: handleEnterKey,
    },
    [],
  );

  return (
    <div
      onDoubleClick={onDoubleClick}
      onClick={handleClick}
      onContextMenu={handleContextMenuClick}
      ref={rootWrapperRef}
      className={`group relative flex h-14 flex-row items-center pl-14 pr-5 ${
        selected ? 'bg-primary/10 text-gray-100 dark:bg-primary/20' : 'focus-within:bg-gray-1 hover:bg-gray-1'
      }`}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      {isOpen && selected && (
        <ContextMenu
          item={item}
          menu={menu}
          menuItemsRef={menuItemsRef}
          openedFromRightClick={openedFromRightClick}
          posX={posX}
          posY={posY}
          isContextMenuCutOff={isContextMenuCutOff}
          genericEnterKey={genericEnterKey}
          handleMenuClose={handleMenuClose}
        />
      )}

      <div
        className={`absolute left-5 top-0 flex h-full w-0 flex-row items-center justify-start p-0 opacity-0 focus-within:opacity-100 group-hover:opacity-100 ${
          selected && 'opacity-100'
        }`}
      >
        <Checkbox
          onClick={(e) => {
            e.stopPropagation();
            onSelectedChanged(!selected);
          }}
          checked={selected}
          checkboxDataCy={`driveListItemCheckbox${listIndex}`}
        />
      </div>
      {disableItemCompositionStyles ? (
        <div
          key={0}
          className={`grow-1 relative flex h-full w-full flex-row items-center border-b ${
            selected ? 'border-primary/5' : 'border-gray-5'
          }`}
        >
          {itemComposition[0](item)}
        </div>
      ) : (
        new Array(itemComposition.length).fill(0).map((_col, i) => (
          <div
            key={i}
            className={`relative flex h-full shrink-0 flex-row items-center border-b ${
              selected ? 'border-primary/5' : 'border-gray-5'
            } ${columnsWidth[i]}`}
          >
            {itemComposition[i](item)}
          </div>
        ))
      )}
      <div
        className={`flex h-14 w-12 shrink-0 flex-col items-center justify-center border-b ${
          selected ? 'border-primary/5' : 'border-gray-5'
        }`}
      >
        <button
          ref={menuButtonRef as LegacyRef<HTMLButtonElement>}
          className={`flex h-10 w-10 flex-col items-center justify-center rounded-md opacity-0 outline-none focus-visible:opacity-100 group-hover:opacity-100 ${
            selected
              ? 'text-gray-80 hover:bg-primary/10 focus-visible:bg-primary/10'
              : 'text-gray-60 hover:bg-gray-10 focus-visible:bg-gray-10'
          }`}
          onClick={(e) => {
            handleContextMenuDotsButton(e);
          }}
        >
          <DotsThree size={24} weight="bold" />
        </button>
      </div>
    </div>
  );
};

export default ListItem;
