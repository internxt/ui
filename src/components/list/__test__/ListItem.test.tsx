import React from 'react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, fireEvent, screen } from '@testing-library/react';
import ListItem from '../ListItem';

const defaultItem = { id: 'item1' };
const mockOnSelectedChanged = vi.fn();
const mockOnDoubleClick = vi.fn();
const mockGenericEnterKey = vi.fn();
const mockOnClickContextMenu = vi.fn();
const mockOnClose = vi.fn();
const itemComposition = [(item: typeof defaultItem) => <div>{item.id}</div>];

const renderListItem = (props = {}) =>
  render(
    <ListItem
      item={defaultItem}
      listIndex={0}
      itemComposition={itemComposition}
      selected={false}
      isOpen={false}
      columnsWidth={['w-1/2']}
      onSelectedChanged={mockOnSelectedChanged}
      onDoubleClick={mockOnDoubleClick}
      genericEnterKey={mockGenericEnterKey}
      onClickContextMenu={mockOnClickContextMenu}
      onClose={mockOnClose}
      {...props}
    />,
  );

describe('ListItem', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('should match snapshot', () => {
    const listItem = renderListItem();
    expect(listItem).toMatchSnapshot();
  });

  it('renders the item correctly', () => {
    renderListItem();
    expect(screen.getByText('item1')).toBeInTheDocument();
  });

  it('calls onDoubleClick when double-clicked', () => {
    renderListItem();
    const listItem = screen.getByText('item1').closest('div');
    fireEvent.doubleClick(listItem!);
    expect(mockOnDoubleClick).toHaveBeenCalledTimes(1);
  });

  it('calls onSelectedChanged when checkbox is clicked', () => {
    renderListItem({ selected: false });
    const checkbox = screen.getByRole('checkbox');
    fireEvent.click(checkbox);
    expect(mockOnSelectedChanged).toHaveBeenCalledWith(true);
  });

  it('calls mockOnClickContextMenu on right-click', () => {
    renderListItem({ isOpen: true, selected: true });
    const listItem = screen.getByText('item1').closest('div');
    fireEvent.contextMenu(listItem!);

    expect(mockOnClickContextMenu).toHaveBeenCalled();
  });

  it('closes the context menu when onClose is called', () => {
    renderListItem({ isOpen: true });
    mockOnClose();
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });
});
