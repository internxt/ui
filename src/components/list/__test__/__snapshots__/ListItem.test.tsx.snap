// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListItem > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="group relative flex h-14 flex-row items-center pl-14 pr-5 focus-within:bg-gray-1 hover:bg-gray-1"
      >
        <div
          class="absolute left-5 top-0 flex h-full w-0 flex-row items-center justify-start p-0 opacity-0 focus-within:opacity-100 group-hover:opacity-100 false"
        >
          <label
            class="relative h-5 w-5 rounded focus-within:outline-primary undefined"
          >
            <div
              class="relative flex h-5 w-5 cursor-pointer flex-col items-center justify-center rounded border text-white border-gray-30 hover:border-gray-40"
              data-cy="driveListItemCheckbox0"
            />
            <input
              class="base-checkbox h-0 w-0 appearance-none opacity-0"
              readonly=""
              type="checkbox"
            />
          </label>
        </div>
        <div
          class="relative flex h-full shrink-0 flex-row items-center border-b border-gray-5 w-1/2"
        >
          <div>
            item1
          </div>
        </div>
        <div
          class="flex h-14 w-12 shrink-0 flex-col items-center justify-center border-b border-gray-5"
        >
          <button
            class="flex h-10 w-10 flex-col items-center justify-center rounded-md opacity-0 outline-none focus-visible:opacity-100 group-hover:opacity-100 text-gray-60 hover:bg-gray-10 focus-visible:bg-gray-10"
          >
            <svg
              fill="currentColor"
              height="24"
              viewBox="0 0 256 256"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M144,128a16,16,0,1,1-16-16A16,16,0,0,1,144,128ZM60,112a16,16,0,1,0,16,16A16,16,0,0,0,60,112Zm136,0a16,16,0,1,0,16,16A16,16,0,0,0,196,112Z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="group relative flex h-14 flex-row items-center pl-14 pr-5 focus-within:bg-gray-1 hover:bg-gray-1"
    >
      <div
        class="absolute left-5 top-0 flex h-full w-0 flex-row items-center justify-start p-0 opacity-0 focus-within:opacity-100 group-hover:opacity-100 false"
      >
        <label
          class="relative h-5 w-5 rounded focus-within:outline-primary undefined"
        >
          <div
            class="relative flex h-5 w-5 cursor-pointer flex-col items-center justify-center rounded border text-white border-gray-30 hover:border-gray-40"
            data-cy="driveListItemCheckbox0"
          />
          <input
            class="base-checkbox h-0 w-0 appearance-none opacity-0"
            readonly=""
            type="checkbox"
          />
        </label>
      </div>
      <div
        class="relative flex h-full shrink-0 flex-row items-center border-b border-gray-5 w-1/2"
      >
        <div>
          item1
        </div>
      </div>
      <div
        class="flex h-14 w-12 shrink-0 flex-col items-center justify-center border-b border-gray-5"
      >
        <button
          class="flex h-10 w-10 flex-col items-center justify-center rounded-md opacity-0 outline-none focus-visible:opacity-100 group-hover:opacity-100 text-gray-60 hover:bg-gray-10 focus-visible:bg-gray-10"
        >
          <svg
            fill="currentColor"
            height="24"
            viewBox="0 0 256 256"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M144,128a16,16,0,1,1-16-16A16,16,0,0,1,144,128ZM60,112a16,16,0,1,0,16,16A16,16,0,0,0,60,112Zm136,0a16,16,0,1,0,16,16A16,16,0,0,0,196,112Z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
